### PR TITLE
Always schedule createTransportRunnable in the executor.

### PR DESCRIPTION
**Pull request by zhangkun83**
Because `scheduleConnection()` is run under lock, if we ran
`createTransportRunnable` inline, `savedDelayedTransport.setTransport()`
will be under lock which violates the assumption made in
https://github.com/grpc/grpc-java/issues/1408 that
> there is an implicit rule today that channel layer will not hold any lock while calling into transport, and had caused deadlock.

### CODE REVIEW

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTransportManagerTest.java [hunk 1]
```diff
@@ -187,7 +194,7 @@ public void reconnect() throws Exception {
     // Pick the first transport
     ClientTransport t1 = tm.getTransport(addressGroup);
     assertNotNull(t1);
-    verify(mockTransportFactory).newClientTransport(addr1, authority);
+    verify(mockTransportFactory, timeout(1000)).newClientTransport(addr1, authority);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplTransportManagerTest.java]: If the test is being ignored, why change the timeout?

File: core/src/main/java/io/grpc/internal/TransportSet.java [hunk 2]
```diff
@@ -194,6 +194,7 @@ public void run() {
         boolean savedShutdown;
         synchronized (lock) {
           savedShutdown = shutdown;
+          reconnectTask = null;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: Is this an unrelated bug fix?

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java [hunk 3]
```diff
@@ -196,6 +197,7 @@ public void twoCallsAndGracefulShutdown() {
     verify(mockTransport, timeout(1000)).newStream(same(method), same(headers));
     verify(mockStream).start(streamListenerCaptor.capture());
     verify(mockStream).setCompressor(isA(Compressor.class));
+    verify(mockStream).setMessageCompression(anyBoolean());
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java]: I'm surprised to see this. Why did it get added?

File: core/src/main/java/io/grpc/internal/TransportSet.java (hunk 2 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/TransportSet.java]: It is a bug caught by tests only with this PR.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java (hunk 3 above)
Comment (zhangkun83) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java]: Before the transport is created in-line, thus `call` will call `start()` on `mockStream` directly, and it doesn't call `setMessageCompression()` unless compression is set.
After the change, there is a good chance that `call` calls `start()` on a `DelayedStream`, which when starting the real stream (`mockStream`), will always call `setMessageCompression()`.

After thinking about this, this is really not something that needs to be checked in this test. I may just remove `verifyNoMoreInteractions(mockStream);` in the end of this test.

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTransportManagerTest.java (hunk 1 above)
Comment (zhangkun83) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplTransportManagerTest.java]: Oops. Removed `@Ignore` and fixed the tests for real.

**Comment by zhangkun83** [comment]
@ejona86 PTAL

**Comment by ejona86** [comment]
@zhangkun83 LGTM

**Comment by zhangkun83** [comment]
This change will trigger a deadlock in `InProcessTransport`, despite that it's fixing another deadlock in `InProcessTransport`. The deadlock it causes is:
```
Found one Java-level deadlock:
=============================
"pool-2959-thread-1":
  waiting to lock monitor 0x00007f9a54015298 (object 0x000000066bf36cd0, a io.grpc.internal.DelayedClientTransport$PendingStream),
  which is held by "grpc-timer-0"
"grpc-timer-0":
  waiting to lock monitor 0x00007f9a5401fc08 (object 0x000000066b857a38, a io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessServerStream),
  which is held by "pool-2959-thread-1"

Java stack information for the threads listed above:
===================================================
"pool-2959-thread-1":
	at io.grpc.internal.DelayedStream.setDecompressor(DelayedStream.java:284)
	- waiting to lock <0x000000066bf36cd0> (a io.grpc.internal.DelayedClientTransport$PendingStream)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.headersRead(ClientCallImpl.java:368)
	at io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessServerStream.writeHeaders(InProcessTransport.java:305)
	- locked <0x000000066b857a38> (a io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessServerStream)
	at io.grpc.internal.ServerCallImpl.sendHeaders(ServerCallImpl.java:149)
	at io.grpc.PartialForwardingServerCall.sendHeaders(PartialForwardingServerCall.java:51)
	at io.grpc.ForwardingServerCall.sendHeaders(ForwardingServerCall.java:37)
	at io.grpc.ForwardingServerCall$SimpleForwardingServerCall.sendHeaders(ForwardingServerCall.java:54)
	at io.grpc.testing.TestUtils$1$1.sendHeaders(TestUtils.java:97)
	at io.grpc.stub.ServerCalls$ResponseObserver.onNext(ServerCalls.java:231)
	at io.grpc.testing.integration.TestServiceImpl$ResponseDispatcher.dispatchChunk(TestServiceImpl.java:280)
	- locked <0x000000066b857a60> (a io.grpc.testing.integration.TestServiceImpl$ResponseDispatcher)
	at io.grpc.testing.integration.TestServiceImpl$ResponseDispatcher.access$000(TestServiceImpl.java:208)
	at io.grpc.testing.integration.TestServiceImpl$ResponseDispatcher$1.run(TestServiceImpl.java:221)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:295)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
"grpc-timer-0":
	at io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessServerStream.clientRequested(InProcessTransport.java)
	- waiting to lock <0x000000066b857a38> (a io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessServerStream)
	at io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessServerStream.access$1400(InProcessTransport.java:213)
	at io.grpc.inprocess.InProcessTransport$InProcessStream$InProcessClientStream.request(InProcessTransport.java:384)
	at io.grpc.internal.DelayedStream.startStream(DelayedStream.java:155)
	at io.gernal.DelayedStream.setStream(DelayedStream.java:179)
	- locked <0x000000066bf36cd0> (a io.grpc.internal.DelayedClientTransport$PendingStream)
	at io.grpc.internal.DelayedClientTransport$PendingStream.createRealStream(DelayedClientTransport.java:243)
	at io.grpc.internal.DelayedClientTransport$PendingStream.access$100(DelayedClientTransport.java:233)
	at io.grpc.internal.DelayedClientTransport.setTransportSupplier(DelayedClientTransport.java:221)
	at io.grpc.internal.DelayedClientTransport.setTransport(DelayedClientTransport.java:181)
	at io.grpc.internal.TransportSet$1.run(TransportSet.java:220)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:295)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I will create another PR with a different fix.
